File: DataStructure/1.Stack/Group1/seoyoon/stack_2.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

bool is_vps(const string s)
{
	stack<char> vps;

	for (char ch : s)
	{
		if (ch == '(')
			vps.push('(');
		else if (ch == ')')
		{
			if (vps.empty())
				return false;
			else
				vps.pop();
		}
	}

	if (vps.empty())
		return true;
	else
		return false;

}

int main()
{
	int num;
	string s;

	cin >> num;
	for (int i=0; i < num; i++)
	{
		cin >> s;
		if (is_vps(s))
			cout << "YES" << endl;
		else
			cout << "NO" << endl;
	}

	return 0;
}

File: DataStructure/1.Stack/Group1/yumin/stack2.py
n = int(input())

for i in range(n):
    stack = []
    PS = input()

    for i in PS:
        if i == '(':
            stack.append(i)
        elif i == ')':
            if len(stack) == 0:
                print("NO")
                break
            else:
                stack.pop()

    else:
        if len(stack) == 0:
            print("YES")
        else:
            print("NO")